### PR TITLE
fix: patch mobile recordings that are missing their meta event

### PR DESCRIPTION
**Pull request by pauldambra**
we had a bug where meta event was not being sent from posthog-ios in some circumstances https://github.com/PostHog/posthog-ios/pull/183

this means a few customers have some otherwise playable recordings they can't watch - because the rrweb player hides itself until a meta event has been seen 🤷 

we _should_ be able to consistently grab width and height from the full snapshot that is being sent, so we can create a meta event to make those recordings playable

----

confirmed locally i can still playback web and mobile recordings and a mobile recording i hacked the meta out of

### CODE REVIEW

File: frontend/src/scenes/session-recordings/player/sessionRecordingDataLogic.test.ts [hunk 1]
```diff
@@ -404,4 +406,30 @@ describe('sessionRecordingDataLogic', () => {
             ])
         })
     })
+
+    describe('snapshot parsing', () => {
```
Comment (pauldambra) [inline frontend/src/scenes/session-recordings/player/sessionRecordingDataLogic.test.ts:410]: there's a lot going on in this method and it only has implicit testing so....

File: frontend/src/scenes/session-recordings/player/sessionRecordingDataLogic.ts [hunk 2]
```diff
@@ -63,15 +64,47 @@ function isRecordingSnapshot(x: unknown): x is RecordingSnapshot {
...
+}
+
 export const parseEncodedSnapshots = async (
     items: (RecordingSnapshot | EncodedRecordingSnapshot | string)[],
     sessionId: string,
     // this is only kept so that we can export the untransformed data for debugging
     withMobileTransformer: boolean = true
 ): Promise<RecordingSnapshot[]> => {
-    if (!postHogEEModule) {
+    if (!postHogEEModule && withMobileTransformer) {
```
Comment (pauldambra) [inline frontend/src/scenes/session-recordings/player/sessionRecordingDataLogic.ts:104]: fly-by: we can delay this import even longer

**Review by marandaneto — APPROVED** [review]

File: frontend/src/scenes/session-recordings/player/sessionRecordingDataLogic.ts [hunk 3]
```diff
@@ -63,6 +64,54 @@ function isRecordingSnapshot(x: unknown): x is RecordingSnapshot {
...
+        // then we need to patch the meta event into the snapshot data
+        if (fullSnapshotIndex > -1 && metaIndex === -1) {
+            const fullSnapshot = parsedLines[fullSnapshotIndex] as fullSnapshotEvent & eventWithTime
+            // a full snapshot (particularly from the mobile transformer) has a relatively fixed structure,
+            // but the types exposed by rrweb don't quite cover what we need , so...
+            const mainNode = fullSnapshot.data.node as any
+            const targetNode = mainNode.childNodes[1].childNodes[1].childNodes[0]
+            const { width, height } = targetNode.attributes
+            const metaEvent = {
+                windowId: fullSnapshot.windowId,
```
[the remark this answers is not in the document]
Comment (pauldambra) [reply frontend/src/scenes/session-recordings/player/sessionRecordingDataLogic.ts:88]: ugh, want to get this out to check it really fixes for them
but type system has been really fighting me here
we're very loose 🤔 😓

File: ee/frontend/mobile-replay/parsing.test.ts [hunk 4]
```diff
@@ -0,0 +1,23 @@
+// import {parseEncodedSnapshots} from "scenes/session-recordings/player/sessionRecordingDataLogic";
+// import {encodedWebSnapshotData} from "scenes/session-recordings/player/__mocks__/encoded-snapshot-data";
```
Comment (daibhin) [inline ee/frontend/mobile-replay/parsing.test.ts:2]: 🔪

File: frontend/src/scenes/session-recordings/player/sessionRecordingDataLogic.ts [hunk 5]
```diff
@@ -83,6 +135,10 @@ export const parseEncodedSnapshots = async (
             const snapshotLine = typeof l === 'string' ? (JSON.parse(l) as EncodedRecordingSnapshot) : l
             const snapshotData = isRecordingSnapshot(snapshotLine) ? [snapshotLine] : snapshotLine['data']
 
+            if (!isMobileSnapshots) {
+                isMobileSnapshots = hasAnyWireframes(snapshotData)
```
Comment (daibhin) [inline frontend/src/scenes/session-recordings/player/sessionRecordingDataLogic.ts:139]: Should we account for screenshot mode too?

File: ee/frontend/mobile-replay/parsing.test.ts (hunk 4 above)
Comment (pauldambra) [reply ee/frontend/mobile-replay/parsing.test.ts:2]: doh

File: frontend/src/scenes/session-recordings/player/sessionRecordingDataLogic.ts (hunk 5 above)
Comment (pauldambra) [reply frontend/src/scenes/session-recordings/player/sessionRecordingDataLogic.ts:139]: screenshot mode sends the screenshots as wireframes... 
wireframe mode doesn't send screenshots

🤯 🤣

**Pull request merged**